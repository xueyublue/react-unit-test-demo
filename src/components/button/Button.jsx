import React from "react";

export default function Button({ label }) {
  return <button data-testid="button">{label}</button>;
}
