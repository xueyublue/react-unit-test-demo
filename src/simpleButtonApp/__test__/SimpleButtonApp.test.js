import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import Button from "../SimpleButtonApp";
import renderer from "react-test-renderer";
import { logRoles } from "@testing-library/dom";

beforeEach(cleanup);

test("button has correct initial color", () => {
  const { container } = render(<Button />);

  // print out roles to console
  logRoles(container);

  const colorButton = screen.getByRole("button", { name: /Change to blue/i });
  expect(colorButton).toHaveStyle({ backgroundColor: "red" });
});

test("button turns blue when clicked", () => {
  render(<Button />);
  const colorButton = screen.getByRole("button", { name: /Change to blue/i });

  // click button
  fireEvent.click(colorButton);

  // expect the background color to be blue
  expect(colorButton).toHaveStyle({ backgroundColor: "blue" });
  // expect the text to be "Change to red"
  expect(colorButton).toHaveTextContent(/Change to red/i);
});

test("matches snapshot", () => {
  const tree = renderer.create(<Button />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("initial conditions", () => {
  render(<Button />);
  const colorButton = screen.getByRole("button", /Change to blue/i);
  // check that the button starts out enabled
  expect(colorButton).toBeEnabled(colorButton);
  // check taht the checkbox started out unchecked
  const checkbox = screen.getByRole("checkbox");
  expect(checkbox).not.toBeChecked(checkbox);
});

test("checkbox disables button on first click and enabls on second click", () => {
  render(<Button />);
  const checkbox = screen.getByRole("checkbox");
  const button = screen.getByRole("button");

  fireEvent.click(checkbox);
  expect(button).toBeDisabled();

  fireEvent.click(checkbox);
  expect(button).toBeEnabled();
});
