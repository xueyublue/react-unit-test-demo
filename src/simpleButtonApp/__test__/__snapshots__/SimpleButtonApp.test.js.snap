// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test frontend behaviors matches snapshot 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "height": "100vh",
      "justifyContent": "center",
      "width": "100vw",
    }
  }
>
  <button
    data-testid="colorButton"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "MediumVioletRed",
        "border": "none",
        "color": "white",
        "padding": "10px 10px",
        "textTransform": "uppercase",
        "transition": "0.25s ease-in-out",
        "width": "25%",
      }
    }
  >
    Change to 
    Midnight Blue
  </button>
  <input
    defaultChecked={false}
    id="disable-button-checkbox"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    htmlFor="disable-button-checkbox"
  >
    Disable Button
  </label>
</div>
`;
